rfcs/0032-run-phase-changes-for-better-nix-shell-use.md

```diff
@@ -0,0 +1,180 @@
+---
+feature: run-phase-changes-for-better-nix-shell-use
+start-date: 2018-08-26
+author: @globin
+co-authors: @dezgeg
+related-issues: https://github.com/dezgeg/nixpkgs/tree/better-run-phases
+shepherd-team: Samuel Dionne-Reil, John Ericson, Linus Heckemann
+shepherd-leader: Linus Heckemann
+---
+
+# Summary
+[summary]: #summary
+
+The intent of this proposal is to tweak a bit how the stdenv runs phases during a build to achieve these two goals:
+1. Improve the UX of nix-shell by making it easier to manually run phases inside a nix-shell
+2. Improve the consistency of pre/post hooks by always running such hooks instead of only running them in non-overridden phases
+
+# Motivation
+[motivation]: #motivation
+
+The primary goal of this RFC is to make it easier to run build phases manually inside a nix-shell.
+Currently, it's a bit annoying because to run say, `buildPhase` the only invocation that works correctly in all cases is:
+
+````sh
+eval "${buildPhase:-buildPhase}"
+````
+
+The goal is to be able to replace the above with the obvious:
+````sh
+buildPhase
+````
+
+The secondary goal is to change how hooks (like `preBuild`, `postBuild` etc.) interact with overridden phases.
+For example a derivation `foo-package` doing,
+````nix
+stdenv.mkDerivation {
+    # ...
+    
+    buildPhase = ''
+        ./my-funky-build-script.sh
+    '';
+    
+    # ...
+}
+````
+causes `preBuild` and `postBuild` not to be called anymore.
+In 99% of the cases this isn't a problem, but it can cause hidden annoyances when using `.overrideAttrs`, for instance:
+````
+    foo-package.overrideAttrs (attrs: {
+        postBuild = (attrs.postBuild or "") + ''
+            # whatever
+        '';
+    })
+````
+Which has led to some people adding explicit `runHook preFoo` and `runHook postFoo` calls to a (small) number of packages.
+
+Thus, to counter this inconsistency, this RFC proposes that those hooks will be run even when the phase is overridden.
+
+# Detailed design
+[design]: #detailed-design
+
+All the 'phase' functions in Nixpkgs need to be reworked a bit. Conceptually, the following diff will be applied to each of them:
+````diff
+-buildPhase() {
++defaultBuildPhase() {
+-    runHook preBuild
+-
+     # set to empty if unset
+     : ${makeFlags=}
+ 
+@@ -1008,14 +1104,14 @@ buildPhase() {
+         make ${makefile:+-f $makefile} "${flagsArray[@]}"
+         unset flagsArray
+     fi
+-
+-    runHook postBuild
+ }
++
++buildPhase() {
++    runHook preBuild
++
++    if [ -n "$buildPhase" ]; then
++        eval "$buildPhase"
++    else
++        defaultBuildPhase
++    fi
++
++    runHook postBuild
++}
+ 
```

Comment:
Has this not been implemented yet, or implemented differently?

Reply:
Haven't seen @dezgeg in a while, don't think this ever got implemented.

Reply:
Fwiw, https://github.com/NixOS/nixpkgs/issues/175605 suggests a perhaps less breaking solution that's opt-in. Opting in is in itself messy though, so that's a reason to prefer this RFC if feasible.